sympy/matrices/tests/test_matrices.py

```diff
@@ -222,31 +223,36 @@ def test_power():
         [0, 0, b**n]])
 
     A = Matrix([[1, 0], [1, 7]])
-    assert A._matrix_pow_by_jordan_blocks(3) == A._eval_pow_by_recursion(3)
+    assert A._matrix_pow_by_jordan_blocks(S(3)) == A._eval_pow_by_recursion(3)
     A = Matrix([[2]])
-    assert A**10 == Matrix([[2**10]]) == A._matrix_pow_by_jordan_blocks(10) == \
+    assert A**10 == Matrix([[2**10]]) == A._matrix_pow_by_jordan_blocks(S(10)) == \
         A._eval_pow_by_recursion(10)
 
     # testing a matrix that cannot be jordan blocked issue 11766
     m = Matrix([[3, 0, 0, 0, -3], [0, -3, -3, 0, 3], [0, 3, 0, 3, 0], [0, 0, 3, 0, 3], [3, 0, 0, 3, 0]])
-    raises(MatrixError, lambda: m._matrix_pow_by_jordan_blocks(10))
+    raises(MatrixError, lambda: m._matrix_pow_by_jordan_blocks(S(10)))
 
     # test issue 11964
-    raises(ValueError, lambda: Matrix([[1, 1], [3, 3]])._matrix_pow_by_jordan_blocks(-10))
+    raises(ValueError, lambda: Matrix([[1, 1], [3, 3]])._matrix_pow_by_jordan_blocks(S(-10)))
     A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 0]])  # Nilpotent jordan block size 3
     assert A**10.0 == Matrix([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
     raises(ValueError, lambda: A**2.1)
     raises(ValueError, lambda: A**(S(3)/2))
     A = Matrix([[8, 1], [3, 2]])
     assert A**10.0 == Matrix([[1760744107, 272388050], [817164150, 126415807]])
     A = Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])  # Nilpotent jordan block size 1
-    assert A**10.2 == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
+    assert A**10.0 == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
```

Comment:
Currently this example would pass with `A**10.2` under this PR:
```
In [16]: A = Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])                                                                            

In [17]: A**10.2                                                                                                                  
Out[17]: 
⎡0  0  1⎤
⎢       ⎥
⎢0  0  1⎥
⎢       ⎥
⎣0  0  1⎦

In [18]: A**10.2 == A                                                                                                             
Out[18]: True
```
You've changed it I assume because we decided that it shouldn't be defined or because it wasn't working at some point?

Looking at the Jordan form I see that it only has trivial Jordan blocks:
```
In [19]: A.jordan_form()                                                                                                          
Out[19]: 
⎛⎡1  0  1⎤  ⎡0  0  0⎤⎞
⎜⎢       ⎥  ⎢       ⎥⎟
⎜⎢0  1  1⎥, ⎢0  0  0⎥⎟
⎜⎢       ⎥  ⎢       ⎥⎟
⎝⎣0  0  1⎦  ⎣0  0  1⎦⎠
```
Maybe that's a case where any nonnegative (not necessarily integer) power can be well defined.

We should probably explicitly document somewhere which cases are expected to work.

Reply:
I changed it because it was obviously a typo for this test judging by what surrounds it. And as I mentioned before since floats are finite then they are all rational, for example 10.2 = 2871044762448691 / 281474976710656, so raising matrix A to this power is equivalent to doing it in two steps (A**p)**(1/q) and again this works out the same as just doing it in one step with the Jordan blocks.

Reply:
Yes it does look like a typo.

Note that SymPy has something of an unresolved philosophical problem with Floats and Rationals. In SymPy a Float is considered to be imprecise in a way that doesn't allow it to be a Rational regardless of the numeric "value" that it has. As a result we have things like:
```julia
In [1]: S(10.0).is_integer
Out[1]: False
```

Reply:
Yes I've run into that bug already. Regardless, from a practical point of view if you want to consider all floats imprecise then the result of the matrix computation will be returned to you equally assumed imprecise, but you get an answer at least.

Reply:
You mentioned a couple of times that we should define what it means for a matrix to be raised to a non-integer and specifically irrational power, well I just thought of one inspired by this idea that all numbers that exist are imprecise (extrapolating a little from SymPy Floats but whatever). Assume an invertible matrix A raised to an irrational r. Now assume some rational p/q which approximates it with a sufficiently large q to get the margin of error to where you are metaphysically comfortable with it. Assume the qth root you get is the principal root (which should be the case I am assuming as all power and sqrt operations in SymPy generally return the principal root). Now when you change r a little bit all you wind up doing is changing the p a little bit since you chose a sufficiently large q so that this is within your margin of error. Changing p in this way does not result in not a linear operation on A, it is exponential, but the important thing is that it is continuous, so there is a continuous path from one approximated irrational to another. Taking the limit as q goes to infinity this should give you a continuous power function of matrices to a real, and it exists and is defined.

Reply:
Specifically I think the awkward case is how do you define the non-integer powers of a non-trivial (size>1), non-invertible Jordan block. Consider that for `[[0, 1], [0, 0]]` then no matrix square root exists.

Reply:
The assumption for what I wrote is that the matrix is invertible since otherwise it is not possible to get the qth root as you note.

Reply:
Oh sorry I see what you mean. Yes I think that we can consistently define non-integer, non-positive and even non-real powers powers of invertible matrices by extension of the formula for the integer case as you say.

I've convinced myself of which cases are well defined and it seems to be self-consistent at least:
```julia
In [6]: M=Matrix([[1, 1], [0, 1]])**n                                                                                             

In [7]: M                                                                                                                         
Out[7]: 
⎡1  n⎤
⎢    ⎥
⎣0  1⎦

In [8]: M.subs(n, x) * M.subs(n, y)                                                                                               
Out[8]: 
⎡1  x + y⎤
⎢        ⎥
⎣0    1  ⎦
```
We can also define nonnegative, real powers for a trivial non-invertible Jordan block:
```julia
In [10]: Matrix([[0]])**pi                                                                                                        
Out[10]: [0]
```
For a non-trivial non-invertible Jordan block only nonnegative integer powers can be defined:
```julia
In [1]: Matrix([[0, 1], [0, 0]])**pi                                                                                              
---------------------------------------------------------------------------
NonInvertibleMatrixError
```
I'm happy with that and I can't think of any examples that aren't covered.

Reply:
Well, infinities, but leave that for another PR.